src/fromager/settings.py

```diff
@@ -1,34 +1,92 @@
 import logging
 import pathlib
+import typing
 
 import yaml
+from packaging.utils import NormalizedName, canonicalize_name
 
 from . import overrides
 
 logger = logging.getLogger(__name__)
 
 
+class DownloadSource(typing.NamedTuple):
+    url: str | None = None
+    rename_to: str | None = None
+
+
+class BuildOption(typing.NamedTuple):
+    # scale parallel jobs by available CPU cores
+    # 1: as many parallel jobs as CPU cores
+    # 2: os.cpu_count() // 2
+    cpu_scaling: int | None = None
+    # scale parallel jobs by memory
+    # 2: assume that each parallel job requires 2 GB max memory
+    memory_scaling: int | None = None
+
+
+class _Data(typing.TypedDict):
+    pre_built: dict[str, set[NormalizedName]]
+    download_source: dict[str, DownloadSource]
+    build_option: dict[str, BuildOption]
+
+
 class Settings:
     def __init__(self, data: dict):
-        self._data = data
+        self._data: _Data = {
+            "pre_built": {},
+            "download_source": {},
+            "build_option": {},
+        }
+
+        # variant -> set of canonicalized package names
+        pre_built: dict[str, typing.Sequence[str]] = data.get("pre_built") or {}
+        self._data["pre_built"] = {
+            variant: set(canonicalize_name(p) for p in packages)
+            for variant, packages in pre_built.items()
+        }
```

Comment:
Throwing an error: 

```
  File "/usr/local/lib/python3.11/site-packages/fromager/settings.py", line 44, in __init__
    self._data["pre_built"] = {
                              ^
  File "/usr/local/lib/python3.11/site-packages/fromager/settings.py", line 45, in <dictcomp>
    variant: set(canonicalize_name(p) for p in packages)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not iterable
```

This is probably because "empty" settings are allowed:
```
pre_built:
    variant1:
    variant 2:
```